### DISCUSSION TITLE
How to read revocation certificate from stream?

### DISCUSSION THREAD

**Discussion by IvanPizhenko** [openpgpjs/openpgpjs#1478] [Q&A]
Note: I'm using Typescript.
Warning: I'm quite new to openpgp.js and may ask questions that are obvious for more experienced people.
I'm trying to port an application that uses openpgp.js v4 to v5. 
Function `Key.getRevocationCertificate()` returns either string or stream:
From `openpgp.d.ts`:
```typescript
  public getRevocationCertificate(date?: Date, config?: Config): Promise<MaybeStream<string> | undefined>;
```
Openpgp v4 have had function `stream.readToEnd()` which allowed to read stream completely. v5 doesn't have it. What is equivalent for v5?

**Comment by IvanPizhenko** [comment]
I've found out that library still include `readToEnd()` and it is just needed to properly add it into `openpgp.d.ts` file. I'm still surprised why it is not included by default. But this limit possibility of library auto-update with Dependabot.

**Comment by larabr** [comment]
Hi @IvanPizhenko , v5 no longer exports the stream utilities, you need to install the [web-stream-tools](https://github.com/openpgpjs/web-stream-tools) separately.

**Reply by IvanPizhenko** [reply]
@larabr Do web-stream-tools have Typescript definitions? I'm trying to add it to my project and getting error like "Could not find a declaration file for module '@openpgp/web-stream-tools'".

**Reply by larabr** [reply]
No definitions yet, unfortunately.

**Reply by IvanPizhenko** [reply]
So then it's easier for me to alter locally `openpgp.d.ts` and `openpgp.js` to export `readToEnd()`.

**Reply by larabr** [reply]
I haven't tried this myself, but locally adding the types for web-stream-tools should be possible like so: https://stackoverflow.com/a/53098695/4569627 .
I've opened an issue about adding TS definitions to the stream tools, but I am not sure when we'll get to implement them: https://github.com/openpgpjs/web-stream-tools/issues/10 . If you want to open a PR there with suggested definitions (even for just a subset of functions), you're welcome to do so 👍

**Reply by IvanPizhenko** [reply]
@larabr Unfortunately, I am not a big specialist in writing `.d.ts` files from JS sources, so I unlikely will do it, at least in the nearest time. I'd better convince you to export `readToEnd` and forget about it.

**Reply by IvanPizhenko** [reply]
@larabr Here is what I've tried:
I've created a file called web-stream-tools.d.ts with following content:
```typescript
declare module "@openpgp/web-stream-tools" {
    type Data = Uint8Array | string;

    interface BaseStream<T extends Data> extends AsyncIterable<T> { }

    interface WebStream<T extends Data> extends BaseStream<T> { // copied+simplified version of ReadableStream from lib.dom.d.ts
      readonly locked: boolean; getReader: Function; pipeThrough: Function; pipeTo: Function; tee: Function;
      cancel(reason?: any): Promise<void>;
    }

    interface NodeStream<T extends Data> extends BaseStream<T> { // copied+simplified version of ReadableStream from @types/node/index.d.ts
      readable: boolean; pipe: Function; unpipe: Function; wrap: Function;
      read(size?: number): string | Uint8Array; setEncoding(encoding: string): this; pause(): this; resume(): this;
      isPaused(): boolean; unshift(chunk: string | Uint8Array): void;
    }

    type Stream<T extends Data> = WebStream<T> | NodeStream<T>;

    type MaybeStream<T extends Data> = T | Stream<T>;

    export function readToEnd<T extends Data>(input: MaybeStream<T>, concat?: (list: T[]) => T): Promise<T>;
}
```
and put it into my `node_modules/@openpgp/web-stream-tools/` and edited its package.json to include "types". 


Then I do in my code:

```typescript
import { readToEnd } from '@openpgp/web-stream-tools';
```

And when trying to run it, I'm getting:
```
  Uncaught exception in build/ts/test.js

  /home/ivan/pro/flowcrypt/flowcrypt-ios/Core/node_modules/@openpgp/web-stream-tools/lib/streams.js:1
  import { isNode, isStream, isArrayStream, isUint8Array, concatUint8Array } from './util';
  ^^^^^^

  SyntaxError: Cannot use import statement outside a module

  › SyntaxError: Cannot use import statement outside a module
  › Object.<anonymous> (build/ts/core/pgp-key.js:14:28)

  ✖ build/ts/test.js exited with a non-zero exit code: 1
```

Either I am doing something wrong or `web-stream-tools` have a bug which makes them just useless. Please, if possible, advise how to solve this.

**Reply by IvanPizhenko** [reply]
@larabr I've posted the above as PR https://github.com/openpgpjs/web-stream-tools/pull/11
If this is appropriate, please merge this. As I said above, I unlikely will write the whole set of type definitions, but please accept this contribution.

**Reply by larabr** [reply]
Thanks, I'll have a look at it
